explorer/interpreter/type_checker.cpp

```diff
@@ -436,40 +437,68 @@ auto TypeChecker::FieldTypesImplicitlyConvertible(
   return true;
 }
 
-// Returns all class members from class and its parent classes.
-static auto GetClassHierarchy(const NominalClassType& class_type)
-    -> std::vector<Nonnull<const NominalClassType*>> {
-  Nonnull<const NominalClassType*> curr_class_type = &class_type;
-  std::vector<Nonnull<const NominalClassType*>> all_classes{curr_class_type};
-  while (curr_class_type->base().has_value()) {
-    curr_class_type = curr_class_type->base().value();
-    all_classes.push_back(curr_class_type);
-  }
-  return all_classes;
-}
-
 auto TypeChecker::FieldTypes(const NominalClassType& class_type) const
     -> std::vector<NamedValue> {
   std::vector<NamedValue> field_types;
-  for (const auto class_type : GetClassHierarchy(class_type)) {
-    for (Nonnull<Declaration*> m : class_type->declaration().members()) {
-      switch (m->kind()) {
-        case DeclarationKind::VariableDeclaration: {
-          const auto& var = cast<VariableDeclaration>(*m);
-          Nonnull<const Value*> field_type =
-              Substitute(class_type->bindings(), &var.binding().static_type());
-          field_types.push_back(
-              {.name = var.binding().name(), .value = field_type});
-          break;
-        }
-        default:
-          break;
+  for (Nonnull<Declaration*> m : class_type.declaration().members()) {
+    switch (m->kind()) {
+      case DeclarationKind::VariableDeclaration: {
+        const auto& var = cast<VariableDeclaration>(*m);
+        Nonnull<const Value*> field_type =
+            Substitute(class_type.bindings(), &var.binding().static_type());
+        field_types.push_back(
+            {.name = var.binding().name(), .value = field_type});
+        break;
       }
+      default:
+        break;
     }
   }
   return field_types;
 }
 
+auto TypeChecker::FieldTypesWithBase(const NominalClassType& class_type) const
+    -> std::vector<NamedValue> {
+  auto fields = FieldTypes(class_type);
+  if (class_type.base().has_value()) {
+    auto base_fields = FieldTypesWithBase(*class_type.base().value());
+    fields.emplace_back(
+        NamedValue{.name = NominalClassValue::base_field,
+                   .value = (new StructType(std::move(base_fields)))});
```

Comment:
```suggestion
                   .value = new StructType(std::move(base_fields))});
```

FWIW I'd find this easier to read without the extra parens. Extra parens can also be risky in LLVM, since they're used to silence certain warnings.

Reply:
Agree. Done.